### PR TITLE
No build directives in generated spec parts

### PR DESCRIPTION
**Pull request by ffesti**
Many things need to be known before the build can be started. Make
declaring these sections or directives an error when encountered parsing
the generated Spec parts.
    
Resolves: #2693

### CODE REVIEW

File: build/parsePreamble.c [hunk 1]
```diff
@@ -1248,7 +1257,29 @@ int parsePreamble(rpmSpec spec, int initialPackage)
...
+		    case RPMTAG_EXCLUDEARCH:
+		    case RPMTAG_EXCLUSIVEARCH:
+		    case RPMTAG_EXCLUDEOS:
+		    case RPMTAG_EXCLUSIVEOS:
+		    case RPMTAG_BUILDROOT:
+		    case RPMTAG_BUILDCONFLICTS:
+		    case RPMTAG_BUILDOPTION:
+		    case RPMTAG_BUILDPREREQ:
+		    case RPMTAG_BUILDREQUIRES:
+		    case RPMTAG_BUILDSYSTEM:
```
Comment (pmatilai) [inline build/parsePreamble.c:1275]: Add this info to the struct in preambleList instead, that forces one to consider this issue whenever new tags are added.
You can check for this from findPreambleTag() just as well.

File: build/rpmbuild_internal.h [hunk 2]
```diff
@@ -21,6 +21,11 @@
 #define ALLOWED_CHARS_EVR ALLOWED_CHARS_VERREL "-:"
 #define LEN_AND_STR(_tag) (sizeof(_tag)-1), (_tag)
 
+enum parseStages {
+    SPECFILE,
+    BUILDSYS,
+    GENERATED,
```
[the remark this answers is not in the document]
Comment (ffesti) [reply build/rpmbuild_internal.h]: Done

File: build/parsePreamble.c (hunk 1 above)
Comment (ffesti) [reply build/parsePreamble.c]: Done.

File: build/parseSpec.c [hunk 3]
```diff
@@ -1109,6 +1123,25 @@ static rpmRC parseSpecSection(rpmSpec *specptr, int secondary)
...
+	    case PART_PREP:
+	    case PART_BUILD:
+	    case PART_INSTALL:
+	    case PART_CHECK:
+	    case PART_CLEAN: /* ???? */
+	    case PART_BUILDARCHITECTURES:
+	    case PART_PATCHLIST:
+	    case PART_SOURCELIST:
+	    case PART_BUILDREQUIRES:
+	    case PART_CONF:
```
Comment (pmatilai) [inline build/parseSpec.c:1137]: Not sure if this was in the earlier version or I just missed it, but this info should go to struct PartRec for the same reason as the tags: to force people to consider the matter when adding new.

Comment (ffesti) [reply build/parseSpec.c]: Yes, it was. But I also was confused when seeing it and at first thought the rebase went wrong.

File: build/parseSpec.c [hunk 4]
```diff
@@ -100,6 +101,30 @@ int isPart(const char *line)
...
+static const int partBeforeBuildOnly(int part)
+{
+    const struct PartRec *p;
+
+    for (p = partList; p->token != NULL; p++) {
+	if (p->part == part) {
+	    return p->beforebuildonly;
+	}
+    }
+    return 0;
```
Comment (pmatilai) [inline build/parseSpec.c:125]: I'd combine these two into findPart() that returns a pointer to the struct, whose values you can then just use directly in the caller.

File: build/parsePreamble.c [hunk 5]
```diff
@@ -1060,71 +1069,72 @@ typedef const struct PreambleRec_s {
...
+    {RPMTAG_REMOVEPATHPOSTFIXES,0, 0, 1, 0, LEN_AND_STR("removepathpostfixes")},
+    {RPMTAG_MODULARITYLABEL,	0, 0, 1, 0, LEN_AND_STR("modularitylabel")},
     {0, 0, 0, 0}
 };
 
 /**
  */
 static int findPreambleTag(rpmSpec spec,rpmTagVal * tag,
-		const char ** macro, char * lang)
+		   const char ** macro, char * lang, int * beforebuildonly)
```
[the remark this answers is not in the document]
Comment (ffesti) [reply build/parsePreamble.c]: Done. I looked at the same issue with `partName`, `partPrebuildonly` and `isPart` but there is is much more difficult to return the struct, as the PART_* value is handed througfh all kind of layers. So I keep things the way they are there.

File: build/parsePreamble.c [hunk 6]
```diff
@@ -1060,71 +1069,72 @@ typedef const struct PreambleRec_s {
     int type;
     int deprecated;
     int ismacro;
+    int beforebuildonly;
```
[the remark this answers is not in the document]
Comment (ffesti) [reply build/parsePreamble.c]: `prebuildonly` it is. Not that I am looking for any hills...

File: build/parseSpec.c (hunk 4 above)
Comment (pmatilai) [reply build/parseSpec.c:125]: There's still this one :innocent:

**Comment by pmatilai** [comment]
Oh and for whatever reason, there's some po/ submodule update in here:

> diff --git a/po b/po
index d5cc5d368e..eee506492f 160000
--- a/po
+++ b/po
@@ -1 +1 @@
-Subproject commit d5cc5d368e2cbb639156b3f6ceb824a8815fdeba
+Subproject commit eee506492f5fe2f4fe8940c5e5b088b42167b790

(also visible in the changed files tab)

File: build/parseSpec.c (hunk 4 above)
Comment (ffesti) [reply build/parseSpec.c:125]: As I wrote above this is not quite as simple as it looks and I ma not that interested in refactoring the whole parting code. Yes, it looks wasteful and over complicated but I'd rather not change  a whole bunch of functions just to pass a new pointer around.

Comment (pmatilai) [reply build/parseSpec.c:125]: Oh, I see you mentioned this below. isPart() is a different matter, but .. 

Once you have 

```
static const struct PartRec * getPart(int part);
```

...the check becomes something like:

```
        if (stage == PARSE_GENERATED) {
            const struct PartRec *p = getPart(parsePart);
            if (p && p->prebuildonly) {
                rpmlog(RPMLOG_ERR,
                        _("Section %s is not allowed after build is done!\n"),
                        p->token);
                goto errxit;
            }
        }   
```

...and at that point you realize this whole if belongs into a separate helper function that checks whether the tag was legitimate for use here. parseSpecSection() is big enough as it is.

Comment (ffesti) [reply build/parseSpec.c]: Not sure if the new code is that more readable. But whatever...

**Review by pmatilai — APPROVED** [review]
I actually meant you could just as well inline the whole lookup in what you're calling checkPart() since there are no other users for this in sight, but works for me.

**Pull request merged**